src/gui/tristateaction.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2019  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QWidgetAction>
+
+class QString;
+class QWidget;
+
+// This class shouldn't be used by anyone except TriStateAction
+// It is required here because it is using Qt signals/slots
+class TriStateWidget : public QWidget
+{
+    Q_OBJECT
+    Q_DISABLE_COPY(TriStateWidget)
+
+public:
+    TriStateWidget(const QString &text, QWidget *parent);
+
+    void setCheckState(Qt::CheckState checkState);
+    void setCloseOnTrigger(bool enabled);
+
+signals:
+    void triggered(bool checked) const;
+
+private:
+    QSize minimumSizeHint() const override;
+
+    void paintEvent(QPaintEvent *) override;
+    void mouseReleaseEvent(QMouseEvent *event) override;
+    void keyPressEvent(QKeyEvent *event) override;
+
+    void transitCheckState();
+
+    bool m_closeOnTriggered;
+    Qt::CheckState m_checkState;
+    const QString m_text;
+};
+
+// TriStateWidget is responsible for checkbox state (tri-state) and paint events while
+// TriStateAction will keep in sync with it. This allows connecting with the usual
+// QAction::triggered slot.
+class TriStateAction : public QWidgetAction
+{
+public:
+    TriStateAction(const QString &text, QWidget *parent);
+
+    void setCheckState(Qt::CheckState checkState);
+    void setCloseOnTrigger(bool enabled);
+
+private:
+    // users should use setCheckState() instead
```

Comment:
Why not reimplement `setChecked()`?

Reply:
There are 2 related functions in Qt: `QAction::setChecked(bool)`, `QCheckBox::setCheckState(Qt::CheckState)`. I followed the latter one.

Reply:
> And I think it is better to have only 1 setter function instead of 2, 2 setters might be confusing.

I see no point in deliberately reducing the existing interface. It can be still called through the base class interface. So it is better to override it to be sure that it works correctly.